cpp/src/codegen/time.cc

```diff
@@ -0,0 +1,484 @@
+// Copyright (C) 2017-2018 Dremio Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "boost/date_time/posix_time/posix_time.hpp"
+
+extern "C" {
+
+#include <stdlib.h>
+#include <time.h>
+
+#include "codegen/time_constants.h"
+#include "codegen/time_types.h"
+
+#define MINS_IN_HOUR 60
+#define SECONDS_IN_MINUTE 60
+
+// Expand inner macro for all date types.
+#define DATE_TYPES(INNER) \
+  INNER(date64)           \
+  INNER(timestamp)
+
+// Extract millennium
+#define EXTRACT_MILLENNIUM(TYPE)                                           \
+  EXTERN                                                                   \
+  int64 extractMillennium##_##TYPE(TYPE millis) {                          \
+    time_t tsec = (time_t)MILLIS_TO_SEC(millis);                           \
```

Comment:
these 3 lines are duplicated in every macro below. can you move it to a macro or an inline function ?